File: experiments/2d_visualization/test_gngt_triple_ring.py
"""Test GNG-T (Triangulation) learning on triple ring pattern.

GNG-T uses heuristic triangulation (quadrilateral search + intersection search)
based on Kubota & Satomi (2008) to maintain proper triangle mesh structure.
"""

import argparse
import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
from PIL import Image

# Add paths for imports
sys.path.insert(0, str(Path(__file__).parents[2] / "algorithms" / "gng_t" / "python"))
sys.path.insert(0, str(Path(__file__).parents[2] / "data" / "2d"))

from model import GrowingNeuralGasT, GNGTParams
from sampler import sample_triple_ring


# Triple ring geometry (matching C++ implementation)
TRIPLE_RING_PARAMS = [
    (0.50, 0.23, 0.06, 0.14),  # top center
    (0.27, 0.68, 0.06, 0.14),  # bottom left
    (0.73, 0.68, 0.06, 0.14),  # bottom right
]


def draw_triple_ring_background(ax) -> None:
    """Draw triple ring background using geometric shapes."""
    theta = np.linspace(0, 2 * np.pi, 100)
    for cx, cy, r_inner, r_outer in TRIPLE_RING_PARAMS:
        outer_x = cx + r_outer * np.cos(theta)
        outer_y = cy + r_outer * np.sin(theta)
        inner_x = cx + r_inner * np.cos(theta)
        inner_y = cy + r_inner * np.sin(theta)
        ax.fill(outer_x, outer_y, color="lightblue", alpha=0.3)
        ax.fill(inner_x, inner_y, color="white")


def create_visualization_frame(
    ax,
    points: np.ndarray,
    nodes: np.ndarray,
    edges: list[tuple[int, int]],
    triangles: list[tuple[int, int, int]] | None,
    iteration: int,
    show_triangles: bool = True,
) -> None:
    """Create a single visualization frame."""
    ax.clear()

    # Draw geometric background
    draw_triple_ring_background(ax)

    # Plot sample points
    ax.scatter(points[:, 0], points[:, 1], c="skyblue", s=3, alpha=0.3)

    # Draw triangles (filled)
    if show_triangles and triangles:
        for tri in triangles:
            triangle = plt.Polygon(
                nodes[list(tri)],
                fill=True,
                facecolor="lightgreen",
                edgecolor="green",
                alpha=0.2,
                linewidth=0.5,
            )
            ax.add_patch(triangle)

    # Plot edges
    for i, j in edges:
        ax.plot(
            [nodes[i, 0], nodes[j, 0]],
            [nodes[i, 1], nodes[j, 1]],
            "r-",
            linewidth=1.5,
            alpha=0.7,
        )

    # Plot nodes
    ax.scatter(nodes[:, 0], nodes[:, 1], c="red", s=50, zorder=5)

    ax.set_xlim(0, 1)
    ax.set_ylim(1, 0)
    ax.set_aspect("equal")
    ax.set_title(f"GNG-T (Triangulation) - Iter {iteration} ({len(nodes)} nodes, {len(edges)} edges)")


def run_experiment(
    n_iterations: int = 5000,
    n_frames: int = 50,
    n_samples: int = 1500,
    output_gif: str = "gngt_triple_ring_growth.gif",
    output_png: str = "gngt_triple_ring_final.png",
    seed: int = 42,
    show_triangles: bool = True,
) -> None:
    """Run GNG-T triple ring experiment."""
    # Sample points using mathematical formula
    print(f"Sampling {n_samples} points from triple ring (mathematical)...")
    points = sample_triple_ring(n_samples=n_samples, seed=seed)
    print(f"Sampled {len(points)} points")

    # Setup GNG-T
    params = GNGTParams(
        max_nodes=100,
        lambda_=100,
        eps_b=0.08,
        eps_n=0.008,
        alpha=0.5,
        beta=0.005,
        max_age=100,
    )
    gng_t = GrowingNeuralGasT(n_dim=2, params=params, seed=seed)

    print(f"Training GNG-T for {n_iterations} iterations...")
    print(f"Parameters: lambda={params.lambda_}, eps_b={params.eps_b}, max_age={params.max_age}")

    # Create figure for animation
    fig, ax = plt.subplots(figsize=(8, 8), facecolor="white")
    frames = []
    frame_interval = max(1, n_iterations // n_frames)

    def callback(model, iteration):
        if iteration % 100 == 0:
            print(f"Iteration {iteration}: {model.n_nodes} nodes, {model.n_edges} edges")

        if iteration % frame_interval == 0 or iteration == n_iterations - 1:
            nodes, edges = model.get_graph()
            triangles = model.get_triangles() if show_triangles else None

            create_visualization_frame(
                ax, points, nodes, edges, triangles, iteration, show_triangles
            )
            fig.canvas.draw()

            img = Image.frombuffer(
                "RGBA", fig.canvas.get_width_height(), fig.canvas.buffer_rgba()
            )
            frames.append(img.convert("RGB"))

    # Train
    gng_t.train(points, n_iterations=n_iterations, callback=callback)

    # Final frame
    nodes, edges = gng_t.get_graph()
    triangles = gng_t.get_triangles() if show_triangles else None
    print(f"Iteration {n_iterations - 1}: {gng_t.n_nodes} nodes, {gng_t.n_edges} edges")

    create_visualization_frame(
        ax, points, nodes, edges, triangles, n_iterations, show_triangles
    )
    fig.savefig(output_png, dpi=100, bbox_inches="tight", facecolor="white")
    print(f"Saved final result: {output_png}")

    plt.close(fig)

    # Save GIF
    if frames:
        frames[0].save(
            output_gif,
            save_all=True,
            append_images=frames[1:],
            duration=100,
            loop=0,
        )
        print(f"Saved GIF: {output_gif}")

    print("Done!")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Test GNG-T on triple ring")
    parser.add_argument("--iterations", type=int, default=5000)
    parser.add_argument("--frames", type=int, default=50)
    parser.add_argument("--samples", type=int, default=1500)
    parser.add_argument("--seed", type=int, default=42)
    parser.add_argument("--no-triangles", action="store_true", help="Hide triangle fill")

    args = parser.parse_args()

    run_experiment(
        n_iterations=args.iterations,
        n_frames=args.frames,
        n_samples=args.samples,
        seed=args.seed,
        show_triangles=not args.no_triangles,
    )

File: experiments/2d_visualization/cpp/test_gngt_triple_ring.cpp
/**
 * @file test_gngt_triple_ring.cpp
 * @brief Test GNG-T on triple ring data (C++ implementation)
 *
 * Based on Kubota & Satomi (2008) GNG with Triangulation.
 * Generates triple ring sample data, runs GNG-T training with 50,000 iterations,
 * and outputs results to CSV files for visualization.
 */

#include <chrono>
#include <cmath>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <random>
#include <sstream>
#include <string>
#include <vector>

#include <Eigen/Core>
#include "gng_t.hpp"

namespace fs = std::filesystem;

/**
 * @brief Sample points from triple ring (3 concentric rings).
 */
std::vector<Eigen::Vector2f> sample_triple_ring(
    int n_samples,
    float center_x, float center_y,
    unsigned int seed)
{
    std::mt19937 rng(seed);
    std::uniform_real_distribution<float> angle_dist(0.0f, 2.0f * M_PI);

    std::vector<std::pair<float, float>> rings = {
        {0.15f, 0.20f},
        {0.25f, 0.30f},
        {0.35f, 0.40f},
    };

    std::vector<Eigen::Vector2f> samples;
    samples.reserve(n_samples);

    int samples_per_ring = n_samples / 3;

    for (const auto& [r_inner, r_outer] : rings) {
        std::uniform_real_distribution<float> radius_dist(r_inner, r_outer);
        for (int i = 0; i < samples_per_ring; ++i) {
            float angle = angle_dist(rng);
            float radius = radius_dist(rng);
            float x = center_x + radius * std::cos(angle);
            float y = center_y + radius * std::sin(angle);
            samples.emplace_back(x, y);
        }
    }

    return samples;
}

void save_samples(const std::vector<Eigen::Vector2f>& samples, const std::string& path) {
    std::ofstream file(path);
    file << "x,y\n";
    file << std::fixed << std::setprecision(6);
    for (const auto& p : samples) {
        file << p.x() << "," << p.y() << "\n";
    }
}

void save_graph(const gng_t::GNGT2f& gng, const std::string& nodes_path,
                const std::string& edges_path, const std::string& triangles_path) {
    std::vector<Eigen::Vector2f> nodes;
    std::vector<std::pair<int, int>> edges;
    gng.get_graph(nodes, edges);

    {
        std::ofstream file(nodes_path);
        file << "x,y\n";
        file << std::fixed << std::setprecision(6);
        for (const auto& n : nodes) {
            file << n.x() << "," << n.y() << "\n";
        }
    }

    {
        std::ofstream file(edges_path);
        file << "i,j\n";
        for (const auto& e : edges) {
            file << e.first << "," << e.second << "\n";
        }
    }

    // Save triangles
    {
        auto triangles = gng.get_triangles();
        std::ofstream file(triangles_path);
        file << "i,j,k\n";
        for (const auto& [i, j, k] : triangles) {
            file << i << "," << j << "," << k << "\n";
        }
    }
}

int main(int argc, char* argv[]) {
    // Default parameters (paper: 50,000 iterations)
    std::string output_dir = "gngt_triple_ring_output";
    int n_samples = 1500;
    int n_iterations = 50000;
    unsigned int seed = 42;
    int frame_interval = 500;

    if (argc > 1) output_dir = argv[1];
    if (argc > 2) n_samples = std::stoi(argv[2]);
    if (argc > 3) n_iterations = std::stoi(argv[3]);
    if (argc > 4) seed = static_cast<unsigned int>(std::stoi(argv[4]));

    fs::create_directories(output_dir);

    std::cout << "GNG-T Triple Ring Test (C++)\n";
    std::cout << "============================\n";
    std::cout << "Output directory: " << output_dir << "\n";
    std::cout << "Samples: " << n_samples << "\n";
    std::cout << "Iterations: " << n_iterations << "\n";
    std::cout << "Seed: " << seed << "\n\n";

    // Generate triple ring samples
    std::cout << "Generating samples...\n";
    auto samples = sample_triple_ring(n_samples, 0.5f, 0.5f, seed);
    save_samples(samples, output_dir + "/samples.csv");
    std::cout << "Generated " << samples.size() << " samples\n\n";

    // Setup GNG-T (paper parameters)
    gng_t::GNGTParams params;
    params.max_nodes = 150;
    params.lambda = 100;
    params.eps_b = 0.08f;
    params.eps_n = 0.008f;
    params.alpha = 0.5f;
    params.beta = 0.005f;
    params.max_age = 100;

    std::cout << "GNG-T Parameters:\n";
    std::cout << "  max_nodes: " << params.max_nodes << "\n";
    std::cout << "  lambda: " << params.lambda << "\n";
    std::cout << "  eps_b: " << params.eps_b << "\n";
    std::cout << "  eps_n: " << params.eps_n << "\n";
    std::cout << "  alpha: " << params.alpha << "\n";
    std::cout << "  beta: " << params.beta << "\n";
    std::cout << "  max_age: " << params.max_age << "\n\n";

    gng_t::GNGT2f model(params, seed);
    model.init();

    std::cout << "Training...\n";

    std::vector<int> frame_iterations;

    auto start_time = std::chrono::high_resolution_clock::now();

    auto callback = [&](const gng_t::GNGT2f& gng, int iter) {
        if (iter % frame_interval == 0 || iter == n_iterations - 1) {
            std::stringstream ss;
            ss << std::setfill('0') << std::setw(5) << iter;
            std::string suffix = ss.str();

            save_graph(gng, output_dir + "/nodes_" + suffix + ".csv",
                       output_dir + "/edges_" + suffix + ".csv",
                       output_dir + "/triangles_" + suffix + ".csv");

            frame_iterations.push_back(iter);

            std::cout << "Iteration " << iter << ": "
                      << gng.num_nodes() << " nodes, "
                      << gng.num_edges() << " edges\n";
        }
    };

    model.train(samples, n_iterations, callback);

    auto end_time = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);

    // Save final result
    save_graph(model, output_dir + "/final_nodes.csv",
               output_dir + "/final_edges.csv",
               output_dir + "/final_triangles.csv");

    // Save frame list
    {
        std::ofstream file(output_dir + "/frames.csv");
        file << "iteration\n";
        for (int iter : frame_iterations) {
            file << iter << "\n";
        }
    }

    // Save metadata
    {
        std::ofstream file(output_dir + "/metadata.csv");
        file << "key,value\n";
        file << "algorithm,GNG-T\n";
        file << "n_iterations," << n_iterations << "\n";
        file << "time_ms," << duration.count() << "\n";
        file << "final_nodes," << model.num_nodes() << "\n";
        file << "final_edges," << model.num_edges() << "\n";
    }

    std::cout << "\nTraining complete!\n";
    std::cout << "Time: " << duration.count() << " ms\n";
    std::cout << "Final: " << model.num_nodes() << " nodes, " << model.num_edges() << " edges\n";
    std::cout << "Results saved to: " << output_dir << "\n";

    return 0;
}
